File: 1-Wstep/intro.py
from math import prod
from os import access
from typing import Any
from numbers import Number

# Wprawki w programowaniu w Python'ie
# Staramy się programować starannie, używając gdzie można "Type Hinting" -
#   po to są te importy w nagłówku
# Ale: jeśli student posługuje się starszymi wersjami Pythona,
# to można mu odpuścić typowanie

# to jest funkcja do eksperymentowania w przyszłości
def myFnc(x, y):
    return x * (x + y)


# Python ma listy
# robimy kilka swoich funkcji pomocniczych

def isEmpty(li : list) -> bool:
    return len(li)==0

# pierwszy element listy
def head(li : list) -> Any: 
    return li[0]

# Ogon listy
def tail(li : list) -> list:
    return li[1:]


# rekurencyjna wersja sumy
def mySum(li : list) -> Number :
    if isEmpty(li) : 
      return 0
    else:  
      return head(li) + mySum(tail(li))


# ZADANIE: przerób mySum na wersję indukcji ogonowej

def mySum2(li : list, acc : Number) -> Number :
    if isEmpty(li):
        return acc
    else:
        return mySum2(tail(li), acc + head(li))

# ZADANIE 11
def sum_of_list(li : list) -> Number :
    if isEmpty(li):
        return 0
    else:
        return head(li) + sum_of_list(tail(li))

def product_of_list(li : list) -> Number:
    if isEmpty(li):
        return 1
    return head(li) * product_of_list(tail(li))

def min_from_list(li : list, min : Number) -> Number:
    if isEmpty(li):
        return min
    elif head(li) < min:
        return min_from_list(tail(li), head(li))
    else:
        return min_from_list(tail(li), min)

def max_from_list(li : list, max : Number) -> Number:
    if isEmpty(li):
        return max
    elif head(li) > max:
        return max_from_list(tail(li), head(li))
    else:
        return max_from_list(tail(li), max)

list = [3, 1, 4, 2]
# print(mySum(list))
# print(mySum2(list, 0))

print(sum_of_list(list))
print(product_of_list(list))
print(min_from_list(list, 123))
print(max_from_list(list, -123))

File: 1-Wstep/intro.js
/*
Wprawki w programowanieu w JavaScript
*/


// do ćwiczenia
function myFnc(x,y) {
    return x*(x+y);
}

// JS nie ma list jako typu natywnego
// modelować je będziemy jako tablice

function isEmpty(arr){
    return arr.length==0
}

//pierwszy element listy
function head(arr) {
    return arr[0];
}

//Ogon listy
function tail(arr) {
    return arr.slice(1)
}

//rekurencyjna wersja sumy

function mySum(arr) {
    if (isEmpty(arr)) {return 0;}
    else {return head(arr) + mySum(tail(arr));}
}

// ZADANIE: przerób mySum na wersję indukcji ogonowej     

function mySum2(arr, acc) {
    if (isEmpty(arr)) { return acc; }
    else { return mySum2(tail(arr), acc + head(arr)); }
}


// ZADANIE 8
function fnc8(x, y) {
    return (z => (w => z + 2*w))(x * y);
}

function fnc8_v2(x, y) {
    return (w => x * y + 2 * w);
}


// ZADANIE 11
function sumOfList(arr) {
    if (isEmpty(arr)) return 0;
    return head(arr) + sumOfList(tail(arr));
}

function productOfList(arr) {
    if (isEmpty(arr)) return 1;
    return head(arr) * productOfList(tail(arr));
}

function minFromList(arr, mini) {
    if (isEmpty(arr)) return mini;
    if (head(arr) < mini) {
        return minFromList(tail(arr), head(arr));
    }
    return minFromList(tail(arr), mini);
}

function maxFromList(arr, maxi) {
    if (isEmpty(arr)) return maxi;
    if (head(arr) > maxi) {
        return maxFromList(tail(arr), head(arr));
    }
    return maxFromList(tail(arr), maxi);
}


let array = [3, 1, 4, 2];
console.log(mySum(array))
console.log(mySum2(array, 0))
console.log(fnc8(2, 3)(6))
console.log(fnc8_v2(2, 3)(6))
console.log(sumOfList(array))
console.log(productOfList(array))
console.log(minFromList(array, 123))
console.log(maxFromList(array, -123))
console.log(array)
